events/websocket.js

```diff
@@ -0,0 +1,360 @@
+  debug("Client asked for unknown method: %s, message: %j",
+        message.method, message);
+  this.sendError("Unknown method", {
+    method:   message.method,
+    data:     message
+  }, message.id);
+};
+
+/** Satisfy a binding request */
+Proxy.prototype.bind = function(id, options) {
+  var that = this;
+
+  // Check binding is an object
+  if (!(options instanceof Object)) {
+    return this.sendError("bind: options must be a JSON object", {
+      options:    options
+    }, id);
+  }
+
+  // Exclude other keys
+  var binding = _.pick(options, 'exchange', 'routingKeyPattern');
+
+  // Check that the routing key pattern is a string
+  if (typeof(binding.routingKeyPattern) !== 'string') {
+    return this.sendError("bind: options.routingKeyPattern must be a string", {
+      options:            options,
+      routingKeyPattern:  options.routingKeyPattern
+    }, id);
+  }
+
+  // Check length routing key pattern
+  if (binding.routingKeyPattern.length > 255) {
+    return this.sendError("bind: options.routingKeyPattern is limited to " +
+                          "255 characters", {
+      options:            options,
+      routingKeyPattern:  options.routingKeyPattern
+    }, id);
+  }
+
+  // Bind to with listener
+  return this.listener.bind(binding).then(function() {
+    // Count number of bindings created
+    that.bindings += 1;
+    // Notify client that binding was successful
+    that.send('bound', {
+      options:      binding
+    }, id);
+  }, function(err) {
+    debug("Failed to bind, binding %j, err %s, as JSON: %j",
+          message.binding, err, err, err.stack);
+    that.sendError("bind: Failed to bind w. requested binding", {
+      options:      binding
+    }, id);
+  }).then(null, function(err) {
+    debug("Failed to reply to socket with method 'bound', err: %s, JSON: %j",
+          err, err, err.stack);
+    // At this stage we panic and close
+    that.close();
+  });
+};
+
+/** Respond to a ping frame */
+Proxy.prototype.onPing = function(data, flags) {
+  try {
+    this.socket.pong(data, flags, true);
+  }
+  catch (err) {
+    debug("Failed to send pong: %s, as JSON: %j", err, err, err.stack);
+    // Close, just as a good measure
+    this.close();
+  }
+};
+
+/** Send a ping to check if the connection is alive */
+Proxy.prototype.keepAlive = function() {
+  // If we have reached the maximum number of inactive iteration close
+  // everything and clean-up
+  if (this.inactiveIterations <= MAX_INACTIVE_ITERATIONS * - 1) {
+    this.sendError("Closing due to missing pongs", {
+      missingPongs:  this.inactiveIterations * - 1
+    });
+    return this.close();
+  }
+  // Increment number of inactive iterations
+  this.inactiveIterations = -1;
+  try {
+    this.socket.ping(slugid.v4());
+  }
+  catch (err) {
+    debug("Failed to send ping");
+    this.close();
+  }
+};
+
+/** Handle pongs, so that keepAlive won't close() in next iteration */
+Proxy.prototype.onPong = function(data, flags) {
```

Comment:
Basically, we need to send something to ensure that the TCP connection is still alive... Technically, we don't have to get something as missing acks will close the TCP connection, but it's nice get pongs... Longer comment added...